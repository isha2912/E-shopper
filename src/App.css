.all-products {
  display: "flex";
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  text-align: center;
  flex-wrap: wrap;
  padding: 3rem;
}
