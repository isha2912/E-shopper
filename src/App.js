/* eslint-disable no-unused-vars */
import React, { useEffect, useState } from 'react';
import { BrowserRouter, Switch, Route } from 'react-router-dom';
import axios from 'axios';
import axiosUtil from './utils/api';
import { ThemeContext } from './components/ThemeContext';
import './App.css';
import Header from './components/Header/Header';
import Home from './components/Home/Home';
import Cart from './components/Cart/Cart';
import Checkout from './components/Checkout/Checkout';
import Allorders from './components/Allorders/Allorders';

const App = () => {
  const [cartItems, setCartItems] = useState({});
  const [orders, setOrders] = useState([]);
  const [isLoaded, setIsLoaded] = useState(false);
  const [cartCount, setCartCount] = useState(0);
  const [theme, setTheme] = useState(0);
  const [filteredProducts, setFilteredProducts] = useState({});

  const onIncrement = (id, category) => {
    const newProduct = filteredProducts[category].map((eachProduct) => (
      eachProduct.id === id ? {
        ...eachProduct,
        count: eachProduct.stock > 0 ? eachProduct.count + 1 : eachProduct.count,
        stock: eachProduct.stock > 0 ? eachProduct.stock - 1 : eachProduct.stock,
      } : eachProduct
    ));
    filteredProducts[category] = newProduct;
    setFilteredProducts(filteredProducts);
    const newCount = (filteredProducts[category].find(
      (product) => {
        if (product.id === id) return (product.stock > 0 && product.count > 0); return false;
      },
    )) ? cartCount + 1 : cartCount;
    setCartCount(newCount);
    const newCartItem = filteredProducts[category].filter((product) => product.count > 0);
    cartItems[category] = newCartItem;
    setCartItems(cartItems);
  };

  const onDecrement = (id, category) => {
    const newProduct = filteredProducts[category].map((product) => (product.id === id ? {
      ...product,
      count: product.count > 0 ? product.count - 1 : product.count,
      stock: product.count > 0 ? product.stock + 1 : product.stock,
    } : product));

    const newCount = (filteredProducts[category].find(
      (product) => { if (product.id === id) return product.count > 0; return false; },
    )) ? cartCount - 1 : cartCount;
    setCartCount(newCount);
    filteredProducts[category] = newProduct;
    setFilteredProducts(filteredProducts);
    const newCartItem = filteredProducts[category].filter((product) => product.count > 0);
    cartItems[category] = newCartItem;
    setCartItems(cartItems);
  };
  useEffect(async () => {
    const receivedOrders = await axiosUtil.getOrders();

    setOrders(receivedOrders);
  }, []);

  useEffect(async () => {
    const items = await axiosUtil.getItems();
    const itemsObjects = items;
    console.log(items);
    const filterProducts = axiosUtil.groupByCategory(itemsObjects);
    setIsLoaded(true);
    setFilteredProducts(filterProducts);
  }, []);

  const onCheckout = async () => {
    const finalOrder = [];
    let finOrder = {};
    Object.values(cartItems).forEach((eachCategory) => {
      eachCategory.forEach((eachProduct) => {
        const item = {
          id: eachProduct.id,
          name: eachProduct.name,
          price: eachProduct.price,
          count: eachProduct.count,
          category: eachProduct.category,
        };
        finalOrder.push(item);
      });
      finOrder = { items: finalOrder };
    });

    const response = await axios.post('./orders', finOrder);
    if (response.data) {
      const newAllOrders = [...orders, response.data.data];
      setOrders(newAllOrders);
      setCartItems({});
      setCartCount(0);
      return true;
    }
    return false;
  };
  return (

    <div className="App">

      <BrowserRouter>
        <ThemeContext.Provider value={theme}>
          <Header cartCount={cartCount} />
        </ThemeContext.Provider>
        <button type="button" onClick={() => setTheme(!theme)}> Theme change </button>
        <Switch>

          <Route path="/" exact>
            <div className="all-products">
              <Home
                filteredProducts={filteredProducts}
                onDecrement={onDecrement}
                onIncrement={onIncrement}
              />
            </div>
          </Route>

          <Route path="/cart" exact>
            <Cart cartItems={cartItems} cartCount={cartCount} />
          </Route>
          <Route path="/checkout" exact>
            <Checkout onSubmitForm={onCheckout} />
          </Route>
          <Route path="/allorders">
            <Allorders allOrders={orders} />
          </Route>

        </Switch>
      </BrowserRouter>

    </div>

  );
};

export default App;
