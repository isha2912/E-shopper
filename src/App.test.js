/* eslint-disable no-unused-vars */
import React from 'react';
import {
  screen, render, waitFor, fireEvent,
} from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import axiosUtil from './utils/api';
import {
  mockOrderedOrder, mockOrders, mockPostOrder, mockProducts,
} from './mockData/api';
import App from './App';

describe(App.name, () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  beforeEach(() => {
    jest.spyOn(axiosUtil, 'getItems').mockResolvedValue(mockProducts);
    jest.spyOn(axiosUtil, 'getOrders').mockResolvedValue(mockOrders);
    jest.spyOn(axiosUtil, 'postOrders').mockResolvedValue(mockPostOrder);
  });

  test('Should match the snapshot', async () => {
    const { container } = await waitFor(() => render(
      <BrowserRouter>
        <App />
      </BrowserRouter>,
    ));
    expect(container).toMatchSnapshot();
  });

  test('Should render header component', async () => {
    const spyProd = jest.spyOn(axiosUtil, 'getItems').mockResolvedValue(mockProducts);
    await waitFor(() => render(<BrowserRouter><App /></BrowserRouter>));
    const allProducts = screen.getByText('ALL PRODUCTS');
    await waitFor(() => { fireEvent.click(allProducts); });
  });

  test('should render home-page when clicked on logo', async () => {
    await waitFor(() => render(
      <BrowserRouter>
        <App />
      </BrowserRouter>,
    ));
    const eShopper = screen.getByTestId('img-tag');

    // screen.logTestingPlaygroundURL();

    await waitFor(() => { fireEvent.click(eShopper); });

    expect(document.location.href).toBe('http://localhost/');
    screen.getByText('apple');
  });

  // test('should render AllOrders page when clicked on AllOrders', async () => {
  //   await waitFor(() => render(
  //     <BrowserRouter>
  //       <App />
  //     </BrowserRouter>,
  //   ));
  //   const allProductLink = screen.getByText('ALL PRODUCTS');
  //   await waitFor(() => fireEvent.click(allProductLink));
  //   expect(document.location.href).toBe('http://localhost/allorders');
  // });

  test('should render cart-page when clicked on Cart-Items', async () => {
    await waitFor(() => render(
      <BrowserRouter>
        <App />
      </BrowserRouter>,
    ));
    const cartLink = screen.getByText('CART COUNT:');
    fireEvent.click(cartLink);
    expect(document.location.href).toBe('http://localhost/cart');
    screen.getByTestId('cart');
  });
});
