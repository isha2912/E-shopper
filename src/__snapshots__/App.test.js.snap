// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Should match the snapshot 1`] = `
<div>
  <div
    class="App"
  >
    <nav
      class="nav-bar-light"
    >
      <a
        href="/"
      >
         
        <div
          class="logo"
        >
          <img
            alt="logo"
            data-testid="img-tag"
            src="./assets/download (1).jpeg"
          />
        </div>
      </a>
      <div
        class="nav-right"
      >
        <a
          href="/allorders"
        >
           
          <h5
            class="all-product-route-white"
          >
             ALL PRODUCTS 
          </h5>
           
        </a>
        <div>
          <div
            class="cart-count"
          >
            <a
              href="/cart"
            >
               
              <h5
                class="cart-count-route-white"
              >
                 CART COUNT:
              </h5>
               
            </a>
            <h5
              class="cart-count-route-white"
            >
               
              0
               
            </h5>
          </div>
           
        </div>
      </div>
    </nav>
    <button
      type="button"
    >
       Theme change 
    </button>
    <div
      class="all-products"
    >
      <div>
        <div
          class="row"
        >
          <h1>
            Fruits & Vegatables
          </h1>
          <div
            class="all-products"
          >
            <div
              class="product-container"
            >
              <img
                alt="fruit"
                class="img-product"
                src="assets/banana.png"
              />
              <div
                class="product-name"
                data-testid="card-img"
              >
                apple
              </div>
              <div
                class="product-quantity"
              >
                0
                 
                Kg
              </div>
              <div
                class="product-price"
              >
                MRP
                 
                120
              </div>
              <div
                class="counter"
              >
                <button
                  class="button-decrement"
                  type="button"
                >
                  -
                </button>
                <div
                  class="in-basket"
                >
                  0
                   
                  in Basket
                </div>
                <button
                  class="button-increment"
                  type="button"
                >
                  +
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
